### PR TITLE
Grid - conditional cell styling 

### LINKED ISSUE #1079: Feature request - Conditional cell styling capability
### Story what I'd like to achieve
I'd like to conditionally change cell background color according to another cell value or something. I think this is a advanced story as below.

> https://github.com/IgniteUI/igniteui-angular/wiki/igxGrid-Specification
Story 12: As an end user, I expect styling capabilities allowing me to set the cell background color and the cell content font color to a custom color provided via its hex encoded string.

### Problem
In order to achieve that, only we can do for now is use cell template like this.
```
// Only when booleanColumn's cell value is true, "checked" class is added to the cell template.
<igx-column [field]="'stringColumn'" [dataType]="'string'" [header]="'string1'">
  <div *igxCell="let value; let cell = cell" [class.checked]="cell.row.rowData['booleanColumn']">
    {{ value }}
  </div>
</igx-column>

.checked {
  color: blue;
  font-weight: bold;
  background-color: salmon;
}
```
But, It doesn't work as I expect because the igx-grid-cell has default padding so that background color of  the cell template can not be filled inside the cell.
image

### Current Solution
So I have to remove default padding of the igx-grid-cell and put padding to the cell template inside the cell.
```
<igx-column [field]="'stringColumn'" [dataType]="'string'" [header]="'string2'" [cellClasses]="'noPadding'">
  <div class="padding" *igxCell="let value; let cell = cell" [class.checked]="cell.row.rowData['booleanColumn']">
    {{ value }}
  </div>
</igx-column>

:host::ng-deep {
  // remove padding from the cell
  .noPadding {
    padding: 0;
  }
  // put padding to the cell template inside the cell
  .padding {
    padding: 1rem 1.5rem;
  }
}
```
It works as we expect.
image

But this solution is a little hassle.

Here is a sample what I explained above. You can change cell background dynamically by clicking checkbox.
https://stackblitz.com/edit/igx-grid-cell-style2

### Expect
I'd like a feature that we can attach conditional classes to the cell directly so that we don't have to remove default padding.

For example, `cellClassses` accepts function in addition to string like below.
```
<igx-column [field]="'stringColumn'" [dataType]="'string'" [header]="'string2'" [cellClasses]="conditionalStyleFunc">

// This function takes cell value and its row data, then returns classes as string.
conditionalStyleFunc(cellValue, rowData): string {
  const cellClasses = ['foo'];
  if (cellValue === 'something') {
    cellClasses.push('bar');
  }
  if (rowData['anotherCell'] === true) {
    cellClasses.push('baz');
  }
  // if all the condition matches, "foo bar baz" would be returned.
  return cellClasses.join(' ');
}
```
It's just an idea. But such a capability would enable the developers more easily handle conditional styling not only background color but also many other css props for cells.

I would appreciate your consideration.

### PR DESCRIPTION
**Pull request by SAndreeva**
Closes #1079 .

### CODE REVIEW

File: projects/igniteui-angular/src/lib/test-utils/grid-cell-style-testing.scss [hunk 1]
```diff
@@ -0,0 +1,17 @@
+:host::ng-deep {
+    .test2 {
+        color: black !important;
```
Comment (bkulov) [inline projects/igniteui-angular/src/lib/test-utils/grid-cell-style-testing.scss]: Do we need `!important`?

File: projects/igniteui-angular/src/lib/grid/cell.component.ts [hunk 2]
```diff
@@ -369,10 +369,16 @@ export class IgxGridCellComponent implements OnInit, OnDestroy, AfterViewInit {
...
     @HostBinding('class')
     get styleClasses(): string {
-        const defaultClasses = [
-            'igx-grid__td igx-grid__td--fw',
-            this.column.cellClasses
-        ];
+        const defaultClasses = ['igx-grid__td igx-grid__td--fw'];
+
+        if (this.column.cellClasses) {
+            Object.entries(this.column.cellClasses).forEach(([name, cb]) => {
```
[the remark this answers is not in the document]
Comment (SAndreeva) [reply projects/igniteui-angular/src/lib/grid/cell.component.ts]: The pollyfill was added with this PR:

[the remark this answers is not in the document]
Comment (SAndreeva) [reply projects/igniteui-angular/src/lib/grid/cell.component.ts]: https://github.com/IgniteUI/igniteui-angular/pull/2528. 
We will also add this as a note in the doc/samples.

[the remark this answers is not in the document]
Comment (SAndreeva) [reply projects/igniteui-angular/src/lib/grid/cell.component.ts]: A polyfill was added with this PR https://github.com/IgniteUI/igniteui-angular/pull/2279. We will take care to add a note for it in the documentation.

File: projects/igniteui-angular/src/lib/grid/cell.spec.ts [hunk 3]
```diff
@@ -1288,3 +1305,442 @@ export class CellEditingScrollTestComponent {
...
+            { field: 'ContactTitle', width: 150, cellClasses: this.cellClasses1 },
+            { field: 'Index', width: 150, cellClasses: this.cellClasses },
+            { field: 'Address', width: 150, cellClasses: this.cellClasses1 },
+            { field: 'City', width: 150, cellClasses: this.cellClasses1 },
+            { field: 'Region', width: 150, cellClasses: this.cellClasses1 },
+            { field: 'PostalCode', width: 150, cellClasses: this.cellClasses1 },
+            { field: 'Phone', width: 150, cellClasses: this.cellClasses1 },
+            { field: 'Fax', width: 150, cellClasses: this.cellClasses1 }
+        ];
+        this.data = [
```
[the remark this answers is not in the document]
Comment (SAndreeva) [reply projects/igniteui-angular/src/lib/grid/cell.spec.ts]: `contactInfoData()` returns a slightly different data. I will simply add one more method to expose mine

[the remark this answers is not in the document]
Comment (SAndreeva) [reply projects/igniteui-angular/src/lib/grid/cell.spec.ts]: ...or find another method.

File: projects/igniteui-angular/src/lib/test-utils/grid-cell-style-testing.scss (hunk 1 above)
Comment (SAndreeva) [reply projects/igniteui-angular/src/lib/test-utils/grid-cell-style-testing.scss]: Yes, when a cell fits two or more conditions, we may want to define which would take precedence...not of great importance, but a possible use case.

**Pull request merged**